dali/pipeline/operators/util/external_source.cu

```diff
@@ -12,21 +12,31 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <utility>
+#include <memory>
+
 #include "dali/pipeline/operators/util/external_source.h"
 
 namespace dali {
 
 template<>
 void ExternalSource<GPUBackend>::RunImpl(DeviceWorkspace *ws) {
-  DALI_ENFORCE(data_in_tl_, "Cannot feed non-contiguous data to GPU op.");
+  std::unique_ptr<TensorList<CPUBackend>> data;
+  {
+    std::unique_lock<std::mutex> busy_lock(busy_m_);
+    cv_.wait(busy_lock, [&data = data_in_tl_]{return !data.empty();});
+    auto is_data_in_tl = data_in_tl_.front();
+    data_in_tl_.pop_front();
+    DALI_ENFORCE(is_data_in_tl, "Cannot feed non-contiguous data to GPU op.");
+    data = tl_data_.PopFront();
+  }
 
   auto &output = ws->Output<GPUBackend>(0);
-  output.Copy(tl_data_, (ws->has_stream() ? ws->stream() : 0));
+  output.Copy(*data, (ws->has_stream() ? ws->stream() : 0));
   {
     std::lock_guard<std::mutex> busy_lock(busy_m_);
-    busy_ = false;
+    tl_data_.Recycle(std::move(data));
```

Comment:
Can you Recycle the buffer without waiting for the copy to finish?

Reply:
I don't think so. It may be possible that after recycle, the user will call SetDataSource and get this recycled tensor before the copy is done.
`Recycle` can be called only when buffer is consumed.

Reply:
I think we had and still have a bug here. As far as I know `output.Copy` is async, so this call is short, and after it ends it doesn't mean that the buffer can be recycled. I think we need to add some callback that will recycle this buffer when copy has been done.

Reply:
That was what I'm asking about.  Recycling it is not safe

Reply:
Previous flow with setting `busy_ = false;` was also wrong. Will fix.